packages/salesforcedx-vscode-apex/src/languageUtils/languageClientUtils.ts

```diff
@@ -104,3 +119,137 @@ export const getExceptionBreakpointInfo = async (): Promise<{}> => {
+    // Server is stopped automatically as part of that procedure
+    try {
+      //if we try and restart during another restart or with a startFailed status, we'll throw an exception
+      await alc.stop();
+    } catch (error) {
+      //if the client is in the state of startFailed, we need to kill the LSP jar and then move on.
+      //todo: do we want to kill LSP jar manually here?
+      vscode.window.showWarningMessage(
+        nls.localize('apex_language_server_restart_dialog_restart_only') + error.message
+      );
+    }
+    if (removeIndexFiles === nls.localize('apex_language_server_restart_dialog_clean_and_restart')) {
+      //delete the apex DB file if we need a clean slate
+      await removeApexDB();
+    }
+    //Need a half second timeout for the creation of a new language client so that we give the previous LS a moment to complete shutdown
+    setTimeout(createLanguageClient, 500, extensionContext, statusBarInstance);
+  }
+};
+
+const removeApexDB = async () => {
+  if (vscode.workspace.workspaceFolders) {
+    const wsrf = vscode.workspace.workspaceFolders[0].uri;
+    const toolsUri = vscode.Uri.joinPath(wsrf, '.sfdx', 'tools');
+    try {
+      const entries = await vscode.workspace.fs.readDirectory(toolsUri);
+      const releaseFolders = entries
+        .filter(([name, type]) => type === vscode.FileType.Directory && /^\d{3}$/.test(name))
+        .map(([name]) => name);
+      for (const folder of releaseFolders) {
+        const folderUri = vscode.Uri.joinPath(toolsUri, folder);
+        await vscode.workspace.fs.delete(folderUri, { recursive: true, useTrash: true });
+      }
+    } catch (error) {
+      console.log('Error, failed to delete folder:' + error.message);
+    }
+  }
+};
+
+export const createLanguageClient = async (
+  extensionContext: vscode.ExtensionContext,
+  languageServerStatusBarItem: ApexLSPStatusBarItem
+): Promise<void> => {
+  const telemetryService = await getTelemetryService();
+  // Initialize Apex language server
+  try {
+    const langClientHRStart = process.hrtime();
+    languageClientUtils.setClientInstance(await languageServer.createLanguageServer(extensionContext));
+
+    const languageClient = languageClientUtils.getClientInstance();
+
+    if (languageClient) {
+      languageClient.errorHandler?.addListener('error', (message: string) => {
+        languageServerStatusBarItem.error(message);
+      });
+      languageClient.errorHandler?.addListener('restarting', (count: number) => {
+        languageServerStatusBarItem.error(
+          nls.localize('apex_language_server_quit_and_restarting').replace('$N', `${count}`)
+        );
+      });
+      languageClient.errorHandler?.addListener('startFailed', () => {
+        languageServerStatusBarItem.error(nls.localize('apex_language_server_failed_activate'));
+      });
+
+      // TODO: the client should not be undefined. We should refactor the code to
```

Comment:
yeah, I agree with this TODO.  you could  do the errors and return if no languageClient and then you don't have the giant conditonal block.